x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java

```diff
@@ -426,13 +429,25 @@ private <T extends TransportResponse> void sendWithCrossClusterAccessHeaders(
             ) {
                 final ThreadContext threadContext = securityContext.getThreadContext();
                 final var contextRestoreHandler = new ContextRestoreResponseHandler<>(threadContext.newRestorableContext(true), handler);
-                try (ThreadContext.StoredContext ignored = threadContext.stashContext()) {
-                    crossClusterAccessHeaders.writeToContext(threadContext);
+                try (ThreadContext.StoredContext ignored = stashContextWithAdditionalHeaders(threadContext, crossClusterAccessHeaders)) {
                     sender.sendRequest(connection, action, request, options, contextRestoreHandler);
                 } catch (Exception e) {
                     contextRestoreHandler.handleException(new SendRequestTransportException(connection.getNode(), action, e));
                 }
             }
+
+            private static ThreadContext.StoredContext stashContextWithAdditionalHeaders(
+                final ThreadContext threadContext,
+                final CrossClusterAccessHeaders crossClusterAccessHeaders
+            ) throws IOException {
+                final String requestId = AuditUtil.extractRequestId(threadContext);
```

Comment:
The alternative here was to introduce an `AuditUtil.stashWithRequestId(threadContext)` method; however, if we ever need to persist additional headers for CCS, the approach I'm taking here makes it easier. I don't feel strongly on this. 

Neither approach helps us keep  `AuditUtil.AUDIT_REQUEST_ID` private unfortunately since we need to allowlist it in `CrossClusterAccessServerTransportFilter`.

Reply:
I'd suggest to move this method into `ThreadContext` and make it more generic, e.g. `ThreadContext#stashContextAndPreserveHeaders(List<String> headerNames)`.

Reply:
I'm good with your suggestion: by making it more generic though we should consider edge cases like passing headers that are already preserved by default etc, whereas in the current implementation it's not widely exposed and so needs to be less stringent.

Reply:
> we should consider edge cases like passing headers that are already preserved by default etc

Just make sure I understand it correctly: In this case, it will be a no-op since it is already preserved, right?

Reply:
Yup I think no-op makes sense. I'll raise it as a separate PR to keep general improvements separate from RCS 2.0 work.

Reply:
IIUC, once we incorporate changes from https://github.com/elastic/elasticsearch/pull/94680, most changes to AuditUtil can be reverted (other than the public constant change)?

Reply:
Yup!